### PR TITLE
Add a subsection on applying a proposal list

### LINKED ISSUE #665: Consolidate Proposal application into its own subsection
* ~Specify inputs and outputs for creation~
    * ~Inputs: Group state, proposals, wire format (pt/ct)~
    * ~Outputs: New group state, MLSMessageContentAuth(Commit), Welcome~
* ~Clarify that the “Provisional” GroupContext includes the extensions specified in any GroupContextExtensions proposal, if any~
* Expand the rules on applying proposals to clearly address how each proposal type is processed
    * Maybe pull into a separate subsection?  “Applying Proposals” after "External Proposals"
* In Commit processing, remove text on applying proposals and just refer to creation steps

(The ~strikethrough~ items above have been fixed by more recent PRs.  Renaming for clearer focus.)

### PR DESCRIPTION
**Pull request by bifurcation**
Closes #665 

Following the pattern of #675, this PR introduces a new section describing how a list of proposals is applied to make a new RatchetTree and GroupContext.  It also moves the validation section introduced in that PR forward so that the sections on validating and applying proposal lists are within the Proposals subsection.

### CODE REVIEW

**Comment by bifurcation** [comment]
One thing I'm not sure about, would like feedback on: When should GroupContextExtensions be applied?  It seems like you could do it either at the front or the back.  I put it at the front so that it could set context for the remaining proposals.  For example, if the GroupContextExtensions introduces a `required_capabilities` extension, then the Adds would be evaluated against it.

**Comment by tomleavy** [comment]
Yes agreed, you sort of need to apply it first so that you can use the RequiredCapabilities to check the other proposals

**Comment by Bren2010** [comment]
Fwiw all that's in this PR currently is moving the validation section

**Comment by bifurcation** [comment]
Thanks @Bren2010, forgot to push.  Latest commit should have all the changes.

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -3553,6 +3553,99 @@ struct {
...
+  `path` field of the external commit MUST meet the same criteria as would the LeafNode
+  in an Update for the removed leaf (see {{update}}). In particular, the `credential`
+  in the LeafNode MUST present a set of identifiers that is acceptable to the
+  application for the removed participant.
+
+* Zero or more PreSharedKey proposals.
+
+* No other proposals.
+
+Proposal type defined in the future may make updates to the above validation
```
Comment (kkohbrok) [inline draft-ietf-mls-protocol.md:3617]:
```suggestion
Proposal types defined in the future may make updates to the above validation
```

File: draft-ietf-mls-protocol.md [hunk 2]
```diff
@@ -3553,6 +3553,99 @@ struct {
...
+logic to incorporate considerations related to proposals of the new type.
+
+## Applying a Proposal List
+
+The sections above defining each proposal type describe how each individual
+proposals are applied.  When creating or processing a Commit, a client applies a
```
Comment (kkohbrok) [inline draft-ietf-mls-protocol.md:3623]: Either leave out the "each" here, or make it "... proposal is ...".

File: draft-ietf-mls-protocol.md [hunk 3]
```diff
@@ -3553,6 +3553,99 @@ struct {
...
+list of proposals to the ratchet tree and GroupContext. The client MUST apply
+the proposals in the list in the following order:
+
+* If there is a GroupContextExtensions proposal, replace the `extensions` field
+  of the GroupContext for the group with the contents of the extension.
+
+* Apply any Update proposals to the ratchet tree, in any order.
+
+* Apply any Remove proposals to the ratceht tree, in any order.
```
Comment (kkohbrok) [inline draft-ietf-mls-protocol.md:3632]:
```suggestion
* Apply any Remove proposals to the ratchet tree, in any order.
```

File: draft-ietf-mls-protocol.md [hunk 4]
```diff
@@ -3553,6 +3553,99 @@ struct {
...
+
+* Apply any Add proposals to the ratchet tree, in the order they appear in the list.
+
+* Look up the PSK secrets for any PreSharedKey proposals, in the order they
+  appear in the list.  These secrets are then used to advance the key schedule
+  later in Commit processing.
+
+* If there is an ExternalInit proposal, use it to derive the `init_secret` for
+  usee later in Commit processing.
```
Comment (kkohbrok) [inline draft-ietf-mls-protocol.md:3641]:
```suggestion
  use later in Commit processing.
```

**Review by kkohbrok — APPROVED** [review]

File: draft-ietf-mls-protocol.md [hunk 5]
```diff
@@ -3553,6 +3553,99 @@ struct {
...
+
+## Applying a Proposal List
+
+The sections above defining each proposal type describe how each individual
+proposals are applied.  When creating or processing a Commit, a client applies a
+list of proposals to the ratchet tree and GroupContext. The client MUST apply
+the proposals in the list in the following order:
+
+* If there is a GroupContextExtensions proposal, replace the `extensions` field
+  of the GroupContext for the group with the contents of the extension.
```
Comment (bifurcation) [inline draft-ietf-mls-protocol.md:3628]: Interim 2022-06-02:

* Clarify that this means that the new extensions apply to other proposals in the same list.

**Review by bifurcation — COMMENTED** [review]
Interim 2022-06-02:

* @bifurcation to resolve comments, then merge

**Pull request merged**